src/number_theory/cyclotomic/primitive_roots.lean

```diff
@@ -0,0 +1,198 @@
+
+/-- If `is_domain B` and `ne_zero (↑n : B)` then `zeta n A B` is a primitive `n`-th root of
+unity. -/
+lemma zeta_primitive_root [is_domain B] [ne_zero ((n : ℕ) : B)] :
+  is_primitive_root (zeta n A B) n :=
+by { rw ←is_root_cyclotomic_iff, exact zeta_spec' n A B }
+
+end is_cyclotomic_extension
+
+end zeta
+
+section no_order
+
+variables [field K] [field L] [comm_ring C] [algebra K L] [algebra K C]
+          [is_cyclotomic_extension {n} K L]
+          {ζ : L} (hζ : is_primitive_root ζ n)
+
+namespace is_primitive_root
+
+/-- The `power_basis` given by `zeta n K L`. -/
+@[simps] noncomputable def power_basis : power_basis K L :=
+power_basis.map (algebra.adjoin.power_basis $ integral {n} K L ζ) $
+(subalgebra.equiv_of_eq _ _ (is_cyclotomic_extension.adjoin_primitive_root_eq_top n _ hζ)).trans
+top_equiv
+
+variables {K}
+
+/-- The equivalence between `L →ₐ[K] A` and `primitive_roots n A` given by a primitive root `ζ`. -/
+@[simps] noncomputable def embeddings_equiv_primitive_roots [is_domain C] [ne_zero ((n : ℕ) : K)]
+  (hirr : irreducible (cyclotomic n K)) : (L →ₐ[K] C) ≃ primitive_roots n C :=
+((hζ.power_basis K).lift_equiv).trans
+{ to_fun    := λ x,
+  by begin
+    haveI hn := ne_zero.of_no_zero_smul_divisors K C n,
+    refine ⟨x.1, _⟩,
+    cases x,
+    rwa [mem_primitive_roots n.pos, ←is_root_cyclotomic_iff, is_root.def,
+         ←map_cyclotomic _ (algebra_map K C), hζ.minpoly_eq_cyclotomic_of_irreducible hirr,
+         ←eval₂_eq_eval_map, ←aeval_def]
+  end,
+  inv_fun   := λ x,
+  by begin
+    haveI hn := ne_zero.of_no_zero_smul_divisors K C n,
+    refine ⟨x.1, _⟩,
+    cases x,
+    rwa [aeval_def, eval₂_eq_eval_map, hζ.power_basis_gen K,
+         ←hζ.minpoly_eq_cyclotomic_of_irreducible hirr, map_cyclotomic, ←is_root.def,
+         is_root_cyclotomic_iff, ← mem_primitive_roots n.pos]
+  end,
+  left_inv  := λ x, subtype.ext rfl,
+  right_inv := λ x, subtype.ext rfl }
+
+end is_primitive_root
+
+namespace is_cyclotomic_extension
+
+variables {K} (L)
+
+/-- If `irreducible (cyclotomic n K)` (in particular for `K = ℚ`), then the `finrank` of a
+cyclotomic extension is `n.totient`. -/
+lemma finrank (hirr : irreducible (cyclotomic n K)) [ne_zero ((n : ℕ) : K)] :
+  finrank K L = (n : ℕ).totient :=
+begin
+  haveI := ne_zero.of_no_zero_smul_divisors K L n,
+  rw [((zeta_primitive_root n K L).power_basis K).finrank, is_primitive_root.power_basis_dim,
+      ←(zeta_primitive_root n K L).minpoly_eq_cyclotomic_of_irreducible hirr, nat_degree_cyclotomic]
+end
+
+end is_cyclotomic_extension
+
+end no_order
+
+namespace is_primitive_root
+
+section norm
+
+variables [linear_ordered_field K] [field L] [comm_ring C] [algebra K L] [algebra K C]
+          {ζ : L} (hζ : is_primitive_root ζ n)
+
+include hζ
+
+/-- If `K` is linearly ordered (in particular for `K = ℚ`), the norm of `ζ` is `1`
+if `n` is odd, for `ζ` a primitive root. -/
+-- TODO: generalize this result to units in general!
```

Comment:
What generalization to units do you have in mind?

Reply:
norm of units is ±1; I did just realise that this won't work fully for us as we probably want exactly 1, but I don't think we have that result yet

Reply:
It's difficult to abstract: the norm of a unit is a unit (this is automatic since `norm` is a bundled multiplicative hom), but in general it will only be a unit of the small field, so just different from `0`. Here what happens is that it a unit, it is integral and the same happens for the inverse, s you get a unit of `Z` aka `±1`, but it is a very specific situation.

Reply:
thanks, I'm not surprised I was missing key details! my lecturers focused too much on ℚ-extensions :b I'll remove this whilst fixing everything else